samza-core/src/main/java/org/apache/samza/zk/BarrierForVersionUpgrade.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.samza.zk;
+
+import java.util.List;
+
+
+/**
+ * Interface for a barrier - to allow synchronization between different processors to switch to a newly published
+ * JobModel.
+ */
+public interface BarrierForVersionUpgrade {
+  /**
+   * Barrier is started by the leader.
+   * @param version - for which the barrier is started.
+   * @param processorsNames - list of processors available at the time of the JobModel generation.
+   */
+  void leaderStartBarrier(String version,  List<String> processorsNames);
```

Comment:
The name of this method is a little weird for me. I would call it startBarrier. If it can only be run by leader, maybe we should put this function in a class Leader belongs to and trigger it with leader.startBarrier().